sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -116,6 +110,71 @@ object GpuProjectExec extends Arm {
       new ColumnarBatch(newColumns, cb.numRows())
     }
   }
+
+  /**
+   * Just like project, but it will try and retry the operations if it can.
+   * @param sb the input batch
+   * @param boundExprs the expressions to run
+   * @return the resulting batch
+   */
+  def projectAndCloseWithRetrySingleBatch(sb: SpillableColumnarBatch,
+      boundExprs: Seq[Expression],
+      spillCallback: SpillCallback): ColumnarBatch = {
+    // First off we want to find/run all of the expressions that are non-deterministic
+    // These cannot be retried.
+    val nonDeterministicExprs = ArrayBuffer[Expression]()
+    val deterministicExprs = ArrayBuffer[Expression]()
+    boundExprs.foreach { expr =>
+      if (expr.deterministic) {
+        deterministicExprs += expr
+      } else {
+        nonDeterministicExprs += expr
+      }
+    }
+
+    val snd = withResource(sb.getColumnarBatch()) { cb =>
+      // First run the nonDeterministic ops because we cannot retry them
+      // In the future we might be able to isolate the non-deterministic expressions and do it more
+      // of a tiered operation
+      if (nonDeterministicExprs.nonEmpty) {
```

Comment:
nit, maybe nice to move this check before we do `sb.getColumnarBatch()`, so we don't need to bring back something from a spilled buffer potentially back to just close it?